File: frontend/src/types/metrics.ts
/**
 * Metrics and billing type definitions
 *
 * Aligns with backend API responses from /api/v1/metrics/*
 */

/**
 * Time series data point for line charts
 */
export interface TimeSeriesDataPoint {
  date: string;
  value: number;
}

/**
 * Token usage data with breakdown
 */
export interface TokenUsageData {
  date: string;
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
}

/**
 * Model usage distribution
 */
export interface ModelDistribution {
  model: string;
  count: number;
  percentage: number;
}

/**
 * Summary statistics for dashboard
 */
export interface MetricStats {
  total_jobs: number;
  total_pages: number;
  total_tokens: number;
  estimated_cost: number;
}

/**
 * Complete dashboard metrics response
 */
export interface DashboardMetrics {
  stats: MetricStats;
  jobs_over_time: TimeSeriesDataPoint[];
  pages_over_time: TimeSeriesDataPoint[];
  tokens_over_time: TokenUsageData[];
  model_distribution: ModelDistribution[];
}

/**
 * Individual completed job for billing
 */
export interface CompletedJob {
  job_id: string;
  document_id: string;
  document_name: string;
  model: string;
  pages_processed: number;
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
  estimated_cost: number;
  processing_time: number;
  completed_at: string;
  tenant_id: string;
}

/**
 * Completed jobs response with pagination
 */
export interface CompletedJobsResponse {
  jobs: CompletedJob[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  total_cost: number;
}

/**
 * Filters for completed jobs query
 */
export interface CompletedJobsFilters {
  start_date?: string;
  end_date?: string;
  model?: string;
}

/**
 * Date range options for metrics
 */
export type DateRangeOption = 7 | 30 | 90;

File: app/schemas/metrics.py
"""Pydantic schemas for metrics API."""
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, ConfigDict, Field


# Dashboard Metrics Schemas
class MetricStats(BaseModel):
    """Summary statistics for dashboard."""

    total_jobs: int = Field(..., description="Total completed jobs")
    total_pages: int = Field(..., description="Total pages processed")
    total_tokens: int = Field(..., description="Total tokens used (input + output)")
    estimated_cost: float = Field(..., description="Estimated total cost in USD")

    model_config = ConfigDict(from_attributes=True)


class TimeSeriesDataPoint(BaseModel):
    """Single data point for time series chart."""

    date: str = Field(..., description="Date in ISO format (YYYY-MM-DD)")
    value: int = Field(..., description="Value for that date")

    model_config = ConfigDict(from_attributes=True)


class TokenUsageData(BaseModel):
    """Token usage breakdown."""

    date: str = Field(..., description="Date in ISO format")
    input_tokens: int = Field(..., description="Input tokens used")
    output_tokens: int = Field(..., description="Output tokens used")
    total_tokens: int = Field(..., description="Total tokens (input + output)")

    model_config = ConfigDict(from_attributes=True)


class ModelDistribution(BaseModel):
    """Model usage distribution for pie chart."""

    model: str = Field(..., description="Model name")
    count: int = Field(..., description="Number of jobs using this model")
    percentage: float = Field(..., description="Percentage of total jobs")

    model_config = ConfigDict(from_attributes=True)


class DashboardMetricsResponse(BaseModel):
    """Response for dashboard metrics endpoint."""

    stats: MetricStats = Field(..., description="Summary statistics")
    jobs_over_time: List[TimeSeriesDataPoint] = Field(
        ..., description="Jobs completed per day"
    )
    pages_over_time: List[TimeSeriesDataPoint] = Field(
        ..., description="Pages processed per day"
    )
    tokens_over_time: List[TokenUsageData] = Field(
        ..., description="Token usage per day"
    )
    model_distribution: List[ModelDistribution] = Field(
        ..., description="Model usage distribution"
    )

    model_config = ConfigDict(from_attributes=True)


# Billing/Completed Jobs Schemas
class CompletedJobItem(BaseModel):
    """Single completed job for billing table."""

    job_id: str = Field(..., description="Extraction job ID")
    document_id: str = Field(..., description="Document ID")
    document_name: Optional[str] = Field(None, description="Original document filename")
    model: str = Field(..., description="Model used for extraction")
    pages_processed: int = Field(..., description="Number of pages processed")
    input_tokens: int = Field(..., description="Input tokens used")
    output_tokens: int = Field(..., description="Output tokens used")
    total_tokens: int = Field(..., description="Total tokens used")
    estimated_cost: float = Field(..., description="Estimated cost in USD")
    processing_time: Optional[float] = Field(
        None, description="Processing time in seconds"
    )
    completed_at: datetime = Field(..., description="Completion timestamp")
    tenant_id: str = Field(..., description="Tenant ID (for admin only)")

    model_config = ConfigDict(from_attributes=True)


class CompletedJobsResponse(BaseModel):
    """Paginated response for completed jobs endpoint."""

    jobs: List[CompletedJobItem] = Field(..., description="List of completed jobs")
    total: int = Field(..., description="Total number of completed jobs")
    page: int = Field(..., description="Current page number (1-indexed)")
    page_size: int = Field(..., description="Number of items per page")
    total_pages: int = Field(..., description="Total number of pages")
    total_cost: float = Field(..., description="Total estimated cost for all jobs")

    model_config = ConfigDict(from_attributes=True)
